sycl/source/exception.cpp

```diff
@@ -10,9 +10,96 @@
 #include <CL/sycl/context.hpp>
 #include <CL/sycl/exception.hpp>
 
+#include <cstring>
+
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 
+namespace { // anonymous
+constexpr char reserved_for_errorcode[] =
+    "01234567812345678"; // 17 (string terminator plus error code)
+std::error_code sycl121_proxy_errorcode = make_error_code(sycl::errc::invalid);
+} // namespace
+
+exception::exception(std::error_code EC, const char *Msg)
+    : exception(EC, nullptr, Msg) {}
+
+exception::exception(std::error_code EC, const std::string &Msg)
+    : exception(EC, nullptr, Msg) {}
+
+// new SYCL 2020 constructors
+exception::exception(std::error_code EC) : exception(EC, nullptr, "") {}
+
+exception::exception(int EV, const std::error_category &ECat,
+                     const std::string &WhatArg)
+    : exception({EV, ECat}, nullptr, WhatArg) {}
+
+exception::exception(int EV, const std::error_category &ECat,
+                     const char *WhatArg)
+    : exception({EV, ECat}, nullptr, std::string(WhatArg)) {}
+
+exception::exception(int EV, const std::error_category &ECat)
+    : exception({EV, ECat}, nullptr, "") {}
+
+exception::exception(context Ctx, std::error_code EC,
+                     const std::string &WhatArg)
+    : exception(EC, std::make_shared<context>(Ctx), WhatArg) {}
+
+exception::exception(context Ctx, std::error_code EC, const char *WhatArg)
+    : exception(Ctx, EC, std::string(WhatArg)) {}
+
+exception::exception(context Ctx, std::error_code EC)
+    : exception(Ctx, EC, "") {}
+
+exception::exception(context Ctx, int EV, const std::error_category &ECat,
+                     const char *WhatArg)
+    : exception(Ctx, {EV, ECat}, std::string(WhatArg)) {}
+
+exception::exception(context Ctx, int EV, const std::error_category &ECat,
+                     const std::string &WhatArg)
+    : exception(Ctx, {EV, ECat}, WhatArg) {}
+
+exception::exception(context Ctx, int EV, const std::error_category &ECat)
+    : exception(Ctx, EV, ECat, "") {}
+
+// protected base constructor for all SYCL 2020 constructors
+exception::exception(std::error_code EC, std::shared_ptr<context> SharedPtrCtx,
+                     const std::string &WhatArg)
+    : MMsg(WhatArg + reserved_for_errorcode), MCLErr(PI_INVALID_VALUE),
+      MContext(SharedPtrCtx) {
+  // For compatibility with previous implementation, we are "hiding" the
+  // std::error_code in the MMsg string, behind the null string terminator
+  int stringTermPoint = MMsg.length() - strlen(reserved_for_errorcode);
```

Comment:
```suggestion
  const int StringTermPoint = MMsg.length() - strlen(ReservedForErrorCode);
```
Please, align style of variable names in the patch.

Reply:
done